include/swift/AST/ArgumentList.h

```diff
@@ -0,0 +1,649 @@
+
+  /// Retrieve an array of the argument labels used in the argument list.
+  ArrayRef<Identifier>
+  getArgumentLabels(SmallVectorImpl<Identifier> &scratch) const;
+
+  /// If the provided expression appears as one of the argument list's
+  /// arguments, returns its index. Otherwise returns \c None. By default this
+  /// will match against semantic sub-expressions, but that may be disabled by
+  /// passing \c false for \c allowSemantic.
+  Optional<unsigned> findArgumentExpr(Expr *expr,
+                                      bool allowSemantic = true) const;
+
+  /// Creates a TupleExpr or ParenExpr that holds the argument exprs. A
+  /// ParenExpr will be returned for a single argument, otherwise a TupleExpr.
+  /// Don't use this function unless you actually need an AST transform.
+  Expr *packIntoImplicitTupleOrParen(
+      ASTContext &ctx,
+      llvm::function_ref<Type(Expr *)> getType = __Expr_getType) const;
+
+  /// Avoid adding new usages of this. Creates a TupleType or ParenType
+  /// representing the types in the argument list. A ParenType will be returned
+  /// for a single argument, otherwise a TupleType.
+  Type composeTupleOrParenType(
+      ASTContext &ctx,
+      llvm::function_ref<Type(Expr *)> getType = __Expr_getType) const;
+
+  /// Whether the argument list matches a given parameter list. This will return
+  /// \c false if the arity doesn't match, or any of the canonical types or
+  /// labels don't match. Note that this expects types to be present for the
+  /// arguments and parameters.
+  bool matches(ArrayRef<AnyFunctionType::Param> params,
+               llvm::function_ref<Type(Expr *)> getType = __Expr_getType) const;
+
+  /// Whether this argument list has the same arguments as another set of
+  /// arguments.
+  bool hasSameArgs(ArrayRef<Argument> args) {
+    if (args.size() != size())
+      return false;
+    return llvm::all_of(indices(args),
+                        [&](auto i) { return get(i) == args[i]; });
+  }
+
+  /// When applying a solution to a constraint system, the type checker rewrites
+  /// argument lists of calls to insert default arguments and collect varargs.
+  /// Sometimes for diagnostics and IDE functionality we want to work on the
+  /// original arguments as written by the user; this performs the reverse
+  /// transformation. This has no effect on un-type-checked argument lists.
+  OriginalArguments getOriginalArguments() const;
+
+  ArgumentList *walk(ASTWalker &walker);
+
+  SWIFT_DEBUG_DUMP;
+  void dump(raw_ostream &OS, unsigned Indent = 0) const;
+};
+
+/// A type that represents the original set of arguments from an ArgumentList.
+///
+/// This always has the same set of arguments as an un-type-checked argument
+/// list. For a type-checked argument list, this strips default arguments,
+/// expands variadic arguments, and re-orders trailing closures such that they
+/// are as they appear at the call-site.
+class OriginalArguments final {
```

Comment:
Typechecking constructs a new `ArgumentList` instead of mutating the original, right? At least in the ways that we would want to undo for `OriginalArguments`?

If so, it might be better to have the typechecked `ArgumentList` contain a pointer to the original `ArgumentList` (perhaps in a trailing object so it can be removed when it's not needed) and then have `getOriginalArguments()` return that pointer if present, or `this` otherwise. That way we wouldn't need two separate types for representing argument lists, and we wouldn't have to reverse-engineer the original arguments from the ones in the AST.

Reply:
Yeah we construct a new `ArgumentList` in CSApply. Storing the un-type-checked argument list as the original did cross my mind when implementing this, but I wasn't sure how it would interact with additional AST transforms done after solver re-writing, and wasn't convinced it was worth the extra word in storage. That being said, storing it in the trailing objects is a clever idea to cut down on that cost, and it may well be that going straight back to the pre-solver-re-written argument list is exactly the behavior we want anyway.

I'll toy around with this and see how it goes! It certainly would be a nice simplification :)

Reply:
Okay, storing the pre-rewritten argument list appears to work just fine and simplifies things nicely, switched over to using that, thanks for the suggestion!